api/tests/mocha/services/setup/view-indexer.spec.js

```diff
@@ -0,0 +1,219 @@
+          json: true,
+          qs: { limit: 1 },
+          timeout: 2000,
+        }],
+      ]);
+    });
+  });
+
+  describe('indexView', () => {
+    it('should query the view with a timeout', async () => {
+      sinon.stub(rpn, 'get').resolves();
+      sinon.stub(env, 'serverUrl').value('http://localhost');
+
+      await viewIndexer.__get__('indexView')('medic', '_design/:staged:medic', 'contacts');
+
+      expect(rpn.get.callCount).to.equal(1);
+      expect(rpn.get.args[0]).to.deep.equal([{
+        uri: 'http://localhost/medic/_design/:staged:medic/_view/contacts',
+        json: true,
+        qs: { limit: 1 },
+        timeout: 2000,
+      }]);
+    });
+
+    it('should retry if the error is a timeout error', async () => {
+      sinon.stub(env, 'serverUrl').value('http://localhost');
+      sinon.stub(rpn, 'get').rejects({ error: { code: 'ESOCKETTIMEDOUT' } });
+      rpn.get.onCall(20).resolves();
+      viewIndexer.__set__('continueIndexing', true);
+
+      await viewIndexer.__get__('indexView')('other', '_design/mydesign', 'viewname');
+
+      expect(rpn.get.callCount).to.equal(21);
+      const params = {
+        uri: 'http://localhost/other/_design/mydesign/_view/viewname',
+        json: true,
+        qs: { limit: 1 },
+        timeout: 2000,
+      };
+      expect(rpn.get.args).to.deep.equal(Array.from({ length: 21 }).map(() => [params]));
+    });
+
+    it('should terminate when other errors are thrown', async () => {
+      sinon.stub(env, 'serverUrl').value('http://localhost');
+      sinon.stub(rpn, 'get').rejects({ error: { code: 'ESOCKETTIMEDOUT' } });
+      rpn.get.onCall(10).rejects({ name: 'error' });
+      viewIndexer.__set__('continueIndexing', true);
+
+      try {
+        await viewIndexer.__get__('indexView')('other', '_design/mydesign', 'viewname');
+        expect.fail('should have thrown');
+      } catch (err) {
+        expect(err).to.deep.equal({ name: 'error' });
+        expect(rpn.get.callCount).to.equal(11);
+      }
+    });
+
+    it('should not throw errors when indexing should stop', async () => {
+      sinon.stub(env, 'serverUrl').value('http://localhost');
+      sinon.stub(rpn, 'get').rejects({ error: { code: 'ESOCKETTIMEDOUT' } });
+      rpn.get.onCall(10).callsFake(() => {
+        viewIndexer.stopIndexing();
+        return Promise.reject({ name: 'error' });
+      });
+      viewIndexer.__set__('continueIndexing', true);
+
+      await viewIndexer.__get__('indexView')('other', '_design/mydesign', 'viewname');
+      expect(rpn.get.callCount).to.equal(11);
+    });
+  });
+
+  describe('stopIndexing', () => {
+    it('should set flag to stop indexing', () => {
+      viewIndexer.__set__('continueIndexing', true);
+      viewIndexer.stopIndexing();
+      expect(viewIndexer.__get__('continueIndexing')).to.equal(false);
+
+      viewIndexer.stopIndexing();
+      expect(viewIndexer.__get__('continueIndexing')).to.equal(false);
+    });
```

Comment:
This is probably unnecessary - it's just testing a setter, right?

I think there's a more version of this test which instead of just checking the internal state, checks that the indexing loop stops after calling `stopIndexing()`.